### ISSUE TITLE
Missed type checking for Objective-C fast dictionary enumeration

### ISSUE THREAD

**Issue by siegel** [llvm/llvm-project#53527]
Consider the following code, which uses type annotations on a dictionary. (It can be compiled standalone using `clang /path/to/file.mm -framework Foundation`.)

```objective-c++
#include <Foundation/Foundation.h>

typedef void (^MyCallbackType)(void);

@interface MyClass : NSObject

@property (nonatomic, strong) NSMutableDictionary<NSUUID*, MyCallbackType> *callbacks;

- (id)registerCallback: (MyCallbackType)myCallback;

- (void)callCallbacks;

@end

@implementation MyClass

- (instancetype)init
{
	if (nil != (self = [super init]))
	{
		self.callbacks = [NSMutableDictionary dictionary];
	}
	
	return self;
}

- (id)registerCallback: (MyCallbackType)callback
{
	NSUUID	*callbackID = NSUUID.UUID;
	
	self.callbacks[callbackID] = Block_copy(callback);
	
	return callbackID;
}

- (void)callCallbacks
{
	//	this is incorrect code, and I think the compiler should report
	//	an error here, since the key type of the dictionary is explicitly `NSUUID`.
	for (MyCallbackType callback in self.callbacks)
		callback();
		
	//	correct formation: use the object enumerator instead
	/*
	for (MyCallbackType callback in self.callbacks.objectEnumerator)
		callback();
	*/
}

@end

int main(int /*argc*/, char **/*argv*/)
{
	MyClass	*foo = [MyClass new];
	
	[foo registerCallback:
		^()
		{
			printf("hello world\n");
		}
	];
	
	[foo callCallbacks];
	
	return 0;
}
```

This may be more "enhancement request" than "bug", but it feels to me as though the compiler should have caught the enumeration type mismatch and issued a diagnostic. As it is, the code will compile, but will (naturally) crash at runtime.
